test/integration/weighted_cluster_integration_test.cc

```diff
@@ -0,0 +1,209 @@
+constexpr int ClusterCount = Envoy::RepickClusterFilter::ClusterCount;
+constexpr int ClusterNameClusterCount = Envoy::RepickClusterFilter::ClusterNameClusterCount;
+
+class WeightedClusterIntegrationTest : public testing::Test, public HttpIntegrationTest {
+public:
+  WeightedClusterIntegrationTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP2, Network::Address::IpVersion::v6) {
+    default_weights_.reserve(ClusterCount);
+    std::fill_n(std::back_inserter(default_weights_), ClusterNameClusterCount, 20);
+    std::fill_n(std::back_inserter(default_weights_),
+                Envoy::RepickClusterFilter::ClusterHeaderClusterCount, 30);
+  }
+
+  void createUpstreams() override {
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+    //  Add fake upstreams
+    for (int i = 0; i < ClusterCount; ++i) {
+      addFakeUpstream(FakeHttpConnection::Type::HTTP2);
+    }
+  }
+
+  void initializeConfig(const std::vector<uint64_t>& weights) {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      // It starts from 1 here because the first cluster is existing cluster configured by
+      // `mutable_clusters` below.
+      for (int i = 1; i < ClusterCount; ++i) {
+        auto* cluster = bootstrap.mutable_static_resources()->add_clusters();
+        cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+        cluster->mutable_load_assignment()
+            ->mutable_endpoints(0)
+            ->mutable_lb_endpoints(0)
+            ->mutable_endpoint()
+            ->mutable_address()
+            ->mutable_socket_address()
+            ->set_address(fake_upstreams_[i]->localAddress()->ip()->addressAsString());
+        cluster->set_name(absl::StrFormat(Envoy::RepickClusterFilter::ClusterNamePrefix, i));
+        ConfigHelper::setHttp2(*cluster);
+      }
+
+      auto* cluster_with_name = bootstrap.mutable_static_resources()->mutable_clusters(0);
+      cluster_with_name->set_name(
+          absl::StrFormat(Envoy::RepickClusterFilter::ClusterNamePrefix, 0));
+      ConfigHelper::setHttp2(*cluster_with_name);
+
+      // Add the custom filter.
+      config_helper_.addFilter("name: repick-cluster-filter");
+    });
+
+    // Modify route with weighted cluster configuration.
+    config_helper_.addConfigModifier(
+        [&weights](
+            envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+                hcm) {
+          // Add the clusters with `name` specified.
+          for (int i = 0; i < ClusterNameClusterCount; ++i) {
+            auto* cluster_with_name = hcm.mutable_route_config()
+                                          ->mutable_virtual_hosts(0)
+                                          ->mutable_routes(0)
+                                          ->mutable_route()
+                                          ->mutable_weighted_clusters()
```

Comment:
You can first store the pointer from the `mutable_weighted_clusters()` and then call `add_cluster()` on it. It will remove duplication of this long chain of calls.